internal/pkg/config/config.go

```diff
@@ -0,0 +1,140 @@
+
+		GlobalConfig: GlobalConfig{
+			GhTokenValidity:       GhTokenValidity,
+			GhTokenUsageWindow:    GhTokenUsageWindow,
+			GhTokenRenewThreshold: GhTokenValidity - GhTokenUsageWindow,
+		},
+	}
+
+}
+
+func LoadConfig(ctx context.Context, client client.Client) *ControllerConfig {
+	log := ctrllog.FromContext(ctx).WithName("ConfigLoader")
+	var configReader struct {
+		Global struct {
+			GhTokenValidity    string `json:"github-token-validity"`
+			GhTokenUsageWindow string `json:"github-token-usage-window"`
+		} `json:"globalconfig"`
+
+		PipelineRun struct {
+			MaxParallelPipelineruns string `json:"max-parallel-pipelineruns"`
+		} `json:"pipelinerunconfig"`
+	}
+
+	defaultConfig := DefaultConfig()
+	config := &ControllerConfig{}
+
+	configMap := &corev1.ConfigMap{}
+	err := client.Get(ctx, types.NamespacedName{
+		Namespace: constant.MintMakerNamespaceName,
+		Name:      ConfigMapName,
+	}, configMap)
+	if err != nil {
+		log.Error(err, "ConfigMap not found, using default configuration", "configMap", ConfigMapName)
+		return defaultConfig
+	}
+
+	if err := json.Unmarshal([]byte(configMap.Data["controller-config"]), &configReader); err != nil {
+		log.Error(err, "Could not unmarshal configuration, using default configuration", "configMap", ConfigMapName)
+		return defaultConfig
+	}
+
+	if parsed, err := strconv.Atoi(configReader.PipelineRun.MaxParallelPipelineruns); err == nil && parsed > 0 {
+		config.PipelineRunConfig.MaxParallelPipelineruns = parsed
+	} else {
+		config.PipelineRunConfig.MaxParallelPipelineruns = defaultConfig.PipelineRunConfig.MaxParallelPipelineruns
+	}
+
+	if parsed, err := time.ParseDuration(configReader.Global.GhTokenValidity); err == nil && parsed > 0 {
+		config.GlobalConfig.GhTokenValidity = parsed
+	} else {
+		config.GlobalConfig.GhTokenValidity = defaultConfig.GlobalConfig.GhTokenValidity
+	}
+
+	if parsed, err := time.ParseDuration(configReader.Global.GhTokenUsageWindow); err == nil && parsed > 0 {
+		config.GlobalConfig.GhTokenUsageWindow = parsed
+	} else {
+		config.GlobalConfig.GhTokenUsageWindow = defaultConfig.GlobalConfig.GhTokenUsageWindow
+	}
+
+	if config.GlobalConfig.GhTokenUsageWindow >= config.GlobalConfig.GhTokenValidity {
+		config.GlobalConfig.GhTokenValidity = defaultConfig.GlobalConfig.GhTokenValidity
+		config.GlobalConfig.GhTokenUsageWindow = defaultConfig.GlobalConfig.GhTokenUsageWindow
+		log.Error(err, "Invalid value for GitHub token usage window, using default",
+			"default", defaultConfig.GlobalConfig.GhTokenUsageWindow)
+	}
+
+	config.GlobalConfig.GhTokenRenewThreshold = config.GlobalConfig.GhTokenValidity - config.GlobalConfig.GhTokenUsageWindow
+
+	return config
+}
+
+// Will not return empty configs but error for logging purposses
+func InitGlobalConfig(ctx context.Context, client client.Client) {
+	config := LoadConfig(ctx, client)
+	globalConfig = config
+}
+
+func GetConfig() *ControllerConfig {
+	return globalConfig
+}
```

Comment:
If you expect it to work as a singleton, it's best to handle concurrency issues.
```
func InitGlobalConfig(ctx context.Context, client client.Client) {
	once.Do(func() {
		config := LoadConfig(ctx, client)
		globalConfig = config
	})
}

func GetConfig() *ControllerConfig {
	if globalConfig == nil {
		return DefaultConfig()
	}
	return globalConfig
}
```